File: packages/shared/src/aoep_shared/voice_profiles.py
"""Narration voice styles for Drive Mode / audio classes.

Maps user preference (or learning-profile signals) to prosody and voice-ranking
hints. Clients pick the best device voice for the locale + style.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Literal, Optional

VoiceStyleId = Literal["standard", "child", "accessible", "calm", "clear"]
NarrationVoicePref = Literal["auto", "standard", "child", "accessible", "calm", "clear"]

VOICE_STYLES: tuple[VoiceStyleId, ...] = (
    "standard",
    "child",
    "accessible",
    "calm",
    "clear",
)

VOICE_STYLE_LABELS: Dict[VoiceStyleId, str] = {
    "standard": "Standard",
    "child": "Child-friendly",
    "accessible": "Accessible (slower)",
    "calm": "Calm & gentle",
    "clear": "Clear & crisp",
}


@dataclass(frozen=True)
class VoiceProsody:
    """Relative speech rate and pitch (1.0 = platform default)."""

    rate: float
    pitch: float


def prosody_for_style(style: VoiceStyleId) -> VoiceProsody:
    if style == "child":
        return VoiceProsody(rate=0.88, pitch=1.12)
    if style == "accessible":
        return VoiceProsody(rate=0.78, pitch=1.0)
    if style == "calm":
        return VoiceProsody(rate=0.85, pitch=0.95)
    if style == "clear":
        return VoiceProsody(rate=0.92, pitch=1.0)
    return VoiceProsody(rate=0.95, pitch=1.0)


def suggest_voice_style_from_profile(profile: Optional[Dict]) -> VoiceStyleId:
    """Infer a narration style from persisted student / learning profile fields."""
    if not profile:
        return "standard"
    age = (profile.get("age_band") or "adult").lower()
    if age == "child":
        return "child"
    acc = profile.get("accessibility") or {}
    pace = (profile.get("learning_pace") or "").lower()
    reading = (profile.get("reading_level") or "").lower()
    if acc.get("needs_extra_time") or pace == "slow" or reading == "beginner":
        return "accessible"
    if acc.get("uses_assistive_tech") or acc.get("needs_captions"):
        return "clear"
    if age == "teen":
        return "clear"
    return "standard"


def resolve_voice_style(
    pref: str,
    profile: Optional[Dict] = None,
) -> VoiceStyleId:
    """Resolve ``auto`` from learning profile; otherwise return the explicit style."""
    if pref and pref != "auto" and pref in VOICE_STYLES:
        return pref  # type: ignore[return-value]
    return suggest_voice_style_from_profile(profile)


def voice_name_style_bonus(style: VoiceStyleId, voice_name: str) -> int:
    """Score adjustment when ranking OS voices by name (web + mobile)."""
    name = (voice_name or "").lower()
    if style == "child":
        if any(k in name for k in ("child", "kids", "junior", "nicky", "pip", "zoe")):
            return 6
        if "compact" in name:
            return 2
    if style == "accessible":
        if any(k in name for k in ("enhanced", "natural", "neural", "premium", "siri")):
            return 4
    if style == "calm":
        if any(k in name for k in ("samantha", "karen", "serena", "moira", "tessa")):
            return 3
    if style == "clear":
        if any(k in name for k in ("enhanced", "natural", "neural", "wavenet", "jenny", "guy")):
            return 5
    if any(k in name for k in ("albert", "zarvox", "whisper", "jester", "bahh")):
        return -8
    return 0

File: apps/web/app/lib/voiceProfiles.ts
// Narration voice styles — mirrors packages/shared aoep_shared.voice_profiles.

import type { StudentProfile } from "./api";

export type NarrationVoiceStyle = "standard" | "child" | "accessible" | "calm" | "clear";
export type NarrationVoicePref = "auto" | NarrationVoiceStyle;

export const NARRATION_VOICE_STYLES: NarrationVoiceStyle[] = [
  "standard", "child", "accessible", "calm", "clear",
];

export const NARRATION_VOICE_LABELS: Record<NarrationVoiceStyle, string> = {
  standard: "Standard",
  child: "Child-friendly",
  accessible: "Accessible (slower)",
  calm: "Calm & gentle",
  clear: "Clear & crisp",
};

export type VoiceProsody = { rate: number; pitch: number };

export function prosodyForStyle(style: NarrationVoiceStyle): VoiceProsody {
  switch (style) {
    case "child": return { rate: 0.88, pitch: 1.12 };
    case "accessible": return { rate: 0.78, pitch: 1.0 };
    case "calm": return { rate: 0.85, pitch: 0.95 };
    case "clear": return { rate: 0.92, pitch: 1.0 };
    default: return { rate: 0.95, pitch: 1.0 };
  }
}

export function suggestVoiceStyleFromProfile(student?: Partial<StudentProfile> | null): NarrationVoiceStyle {
  if (!student) return "standard";
  if (student.age_band === "child") return "child";
  const acc = student.accessibility || {};
  if (acc.needs_extra_time || student.learning_pace === "slow" || student.reading_level === "beginner") {
    return "accessible";
  }
  if (acc.uses_assistive_tech || acc.needs_captions) return "clear";
  if (student.age_band === "teen") return "clear";
  return "standard";
}

export function resolveVoiceStyle(
  pref: NarrationVoicePref,
  student?: Partial<StudentProfile> | null,
): NarrationVoiceStyle {
  if (pref !== "auto") return pref;
  return suggestVoiceStyleFromProfile(student);
}

export function voiceNameStyleBonus(style: NarrationVoiceStyle, voiceName: string): number {
  const name = (voiceName || "").toLowerCase();
  if (style === "child") {
    if (/(child|kids|junior|nicky|pip|zoe)/.test(name)) return 6;
    if (name.includes("compact")) return 2;
  }
  if (style === "accessible") {
    if (/(enhanced|natural|neural|premium|siri)/.test(name)) return 4;
  }
  if (style === "calm") {
    if (/(samantha|karen|serena|moira|tessa)/.test(name)) return 3;
  }
  if (style === "clear") {
    if (/(enhanced|natural|neural|wavenet|jenny|guy)/.test(name)) return 5;
  }
  if (/(albert|zarvox|whisper|jester|bahh)/.test(name)) return -8;
  return 0;
}
